File: app/src/main/java/com/glia/exampleapp/GliaWidgetsConfigManager.java
package com.glia.exampleapp;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceManager;

import com.glia.androidsdk.SiteApiKey;
import com.glia.widgets.GliaWidgetsConfig;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 *Helper class to obtain Glia Config params from deep-link or preferences.
 *
 *for setup with secret link must be
 *      glia://widgets/secret?site_id={site_id}&api_key_secret={api_key_secret}&api_key_id={api_key_id}&queue_id={queue_id}&visitor_context_asset_id={visitor_context_asset_id}
 *      where all query params are mandatory except visitor_context_asset_id
 */
public class GliaWidgetsConfigManager {
    private static final String SECRET_KEY = "secret";

    private static final String SITE_ID_KEY = "site_id";
    private static final String API_KEY_SECRET_KEY = "api_key_secret";
    private static final String API_KEY_ID_KEY = "api_key_id";
    private static final String QUEUE_ID_KEY = "queue_id";
    public static final String VISITOR_CONTEXT_ASSET_ID_KEY = "visitor_context_asset_id";

    private static final String REGION_BETA = "beta";
    private static final String REGION_ACCEPTANCE = "acceptance";

    @NonNull
    public static GliaWidgetsConfig obtainConfigFromDeepLink(@NonNull Uri data, @NonNull Context applicationContext) {
        saveQueueIdToPrefs(data, applicationContext);
        saveVisitorContextAssetIdIfPresent(data, applicationContext);
        saveSiteIdToPrefs(data, applicationContext);

//        We're not checking availability for every query param separately because this is for acceptance tests and we assume that link would be correct
        if (SECRET_KEY.equals(data.getLastPathSegment())) {
            saveSiteApiKeyAuthToPrefs(data, applicationContext);
            return obtainConfigFromSecretDeepLink(data, applicationContext);
        } else {
            throw new RuntimeException("deep link must start with \"glia://widgets/secret\"");
        }
    }

    private static void saveVisitorContextAssetIdIfPresent(Uri data, Context applicationContext) {
        String visitorContextAssetId = data.getQueryParameter(VISITOR_CONTEXT_ASSET_ID_KEY);

        if (visitorContextAssetId == null) return;

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext);

        sharedPreferences.edit().putString(applicationContext.getString(R.string.pref_context_asset_id), visitorContextAssetId).apply();
    }

    private static void saveQueueIdToPrefs(@NonNull Uri data, @NonNull Context applicationContext) {
        String queueId = data.getQueryParameter(QUEUE_ID_KEY);

        if (queueId == null) return;

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext);

        sharedPreferences.edit().putString(applicationContext.getString(R.string.pref_queue_id), queueId).apply();
    }

    private static void saveSiteIdToPrefs(@NonNull Uri data, @NonNull Context applicationContext) {
        String siteId = data.getQueryParameter(SITE_ID_KEY);

        if (siteId == null) return;

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext);

        sharedPreferences.edit().putString(applicationContext.getString(R.string.pref_site_id), siteId).apply();
    }

    private static void saveSiteApiKeyAuthToPrefs(@NonNull Uri data, @NonNull Context applicationContext) {
        String apiKeyId = data.getQueryParameter(API_KEY_ID_KEY);
        String apiKeySecret = data.getQueryParameter(API_KEY_SECRET_KEY);

        if (apiKeyId == null || apiKeySecret == null) return;

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext);

        sharedPreferences.edit()
                .putString(applicationContext.getString(R.string.pref_api_key_id), apiKeyId)
                .putString(applicationContext.getString(R.string.pref_api_key_secret), apiKeySecret)
                .apply();
    }

    @NonNull
    private static GliaWidgetsConfig obtainConfigFromSecretDeepLink(@NonNull Uri data, @NonNull Context applicationContext) {
        return new GliaWidgetsConfig.Builder()
                .setSiteApiKey(new SiteApiKey(data.getQueryParameter(API_KEY_ID_KEY), data.getQueryParameter(API_KEY_SECRET_KEY)))
                .setSiteId(data.getQueryParameter(SITE_ID_KEY))
                .setRegion(REGION_ACCEPTANCE)
                .setContext(applicationContext)
                .build();
    }

    @NonNull
    public static GliaWidgetsConfig createDefaultConfig(@NonNull Context applicationContext) {
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext);
        return configWithSiteApiKeyAuth(sharedPreferences, applicationContext);
    }

    @NonNull
    private static GliaWidgetsConfig configWithSiteApiKeyAuth(@NonNull SharedPreferences preferences, @NonNull Context applicationContext) {
        String apiKeyId = preferences.getString(applicationContext.getString(R.string.pref_api_key_id), applicationContext.getString(R.string.glia_api_key_id));
        String apiKeySecret = preferences.getString(applicationContext.getString(R.string.pref_api_key_secret), applicationContext.getString(R.string.glia_api_key_secret));
        String siteId = preferences.getString(applicationContext.getString(R.string.pref_site_id), applicationContext.getString(R.string.site_id));


        return new GliaWidgetsConfig.Builder()
                .setSiteApiKey(new SiteApiKey(apiKeyId, apiKeySecret))
                .setSiteId(siteId)
                .setRegion(REGION_BETA)
                .setContext(applicationContext)
                .setUiJsonRemoteConfig(getRawResource(applicationContext, R.raw.global_colors)) // TODO MOB-1919
                .build();
    }

    // for testing the remote theme configurations JSON
    private static String getRawResource(Context context, int resource) {
        String res = null;
        InputStream is = context.getResources().openRawResource(resource);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] b = new byte[1];
        try {
            while ( is.read(b) != -1 ) {
                baos.write(b);
            }
            res = baos.toString();
            is.close();
            baos.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return res;
    }

}

File: app/src/main/java/com/glia/exampleapp/GliaWidgetsConfigManager.kt
package com.glia.exampleapp

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.preference.PreferenceManager
import com.glia.androidsdk.SiteApiKey
import com.glia.androidsdk.screensharing.ScreenSharing
import com.glia.widgets.GliaWidgetsConfig
import com.glia.widgets.UiTheme

/**
 * Helper class to obtain Glia Config params from deep-link or preferences.
 *
 * for setup with a secret link must be
 * glia://widgets/secret?site_id={site_id}&api_key_secret={api_key_secret}&api_key_id={api_key_id}&queue_id={queue_id}&visitor_context_asset_id={visitor_context_asset_id}
 * where all query params are mandatory except visitor_context_asset_id
 */
object GliaWidgetsConfigManager {
    private const val SECRET_KEY = "secret"
    private const val SITE_ID_KEY = "site_id"
    private const val API_KEY_SECRET_KEY = "api_key_secret"
    private const val API_KEY_ID_KEY = "api_key_id"
    private const val QUEUE_ID_KEY = "queue_id"
    private const val VISITOR_CONTEXT_ASSET_ID_KEY = "visitor_context_asset_id"
    private const val REGION_KEY = "environment"
    private const val REGION_BETA = "beta"
    private const val REGION_ACCEPTANCE = "acceptance"
    @JvmStatic
    fun obtainConfigFromDeepLink(data: Uri, applicationContext: Context): GliaWidgetsConfig {
        saveQueueIdToPrefs(data, applicationContext)
        saveVisitorContextAssetIdIfPresent(data, applicationContext)
        saveSiteIdToPrefs(data, applicationContext)

//       We're not checking availability for every query param separately because this is for acceptance tests and we assume that link would be correct
        if (SECRET_KEY == data.lastPathSegment) {
            saveSiteApiKeyAuthToPrefs(data, applicationContext)
        } else {
            throw RuntimeException("deep link must start with \"glia://widgets/secret\"")
        }
        val region = data.getQueryParameter(REGION_KEY) ?: REGION_ACCEPTANCE

        // By this point all settings from deep link where saved to the shared prefs overriding the
        // defaults combining both together
        return createDefaultConfig(
            context = applicationContext,
            region = region
        )
    }

    private fun saveVisitorContextAssetIdIfPresent(data: Uri, applicationContext: Context) {
        val visitorContextAssetId = data.getQueryParameter(VISITOR_CONTEXT_ASSET_ID_KEY) ?: return
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        sharedPreferences.edit().putString(
            applicationContext.getString(R.string.pref_context_asset_id),
            visitorContextAssetId
        ).apply()
    }

    private fun saveQueueIdToPrefs(data: Uri, applicationContext: Context) {
        val queueId = data.getQueryParameter(QUEUE_ID_KEY) ?: return
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        sharedPreferences.edit()
            .putString(applicationContext.getString(R.string.pref_queue_id), queueId)
            .apply()
    }

    private fun saveSiteIdToPrefs(data: Uri, applicationContext: Context) {
        val siteId = data.getQueryParameter(SITE_ID_KEY) ?: return
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        sharedPreferences.edit()
            .putString(applicationContext.getString(R.string.pref_site_id), siteId)
            .apply()
    }

    private fun saveSiteApiKeyAuthToPrefs(data: Uri, applicationContext: Context) {
        val apiKeyId = data.getQueryParameter(API_KEY_ID_KEY)
        val apiKeySecret = data.getQueryParameter(API_KEY_SECRET_KEY)
        if (apiKeyId == null || apiKeySecret == null) return
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(applicationContext)
        sharedPreferences.edit()
            .putString(applicationContext.getString(R.string.pref_api_key_id), apiKeyId)
            .putString(applicationContext.getString(R.string.pref_api_key_secret), apiKeySecret)
            .apply()
    }

    @JvmStatic
    @JvmOverloads
    fun createDefaultConfig(
        context: Context,
        uiJsonRemoteConfig: String? = null,
        runtimeConfig: UiTheme? = null,
        region: String = REGION_BETA,
        preferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    ): GliaWidgetsConfig {
        val apiKeyId = preferences.getString(
            context.getString(R.string.pref_api_key_id),
            context.getString(R.string.glia_api_key_id)
        )
        val apiKeySecret = preferences.getString(
            context.getString(R.string.pref_api_key_secret),
            context.getString(R.string.glia_api_key_secret)
        )
        val siteId = preferences.getString(
            context.getString(R.string.pref_site_id),
            context.getString(R.string.site_id)
        )
        val companyName = preferences.getString(context.getString(R.string.pref_company_name), context.getString(R.string.settings_value_default_company_name))
        val useOverlay = preferences.getBoolean(context.getString(R.string.pref_use_overlay), true)
        val bounded = context.getString(R.string.screen_sharing_mode_app_bounded)
        val unbounded = context.getString(R.string.screen_sharing_mode_unbounded)
        val screenSharingMode = if (preferences.getString(
                context.getString(R.string.pref_screen_sharing_mode), unbounded
            ) == bounded
        ) {
            ScreenSharing.Mode.APP_BOUNDED
        } else {
            ScreenSharing.Mode.UNBOUNDED
        }
        return GliaWidgetsConfig.Builder()
            .setSiteApiKey(SiteApiKey(apiKeyId!!, apiKeySecret!!))
            .setSiteId(siteId)
            .setRegion(region)
            .setCompanyName(companyName)
            .setUseOverlay(useOverlay)
            .setScreenSharingMode(screenSharingMode)
            .setContext(context)
            .setUiJsonRemoteConfig(uiJsonRemoteConfig ?: Utils.getRemoteThemeByPrefs(preferences, context.resources))
            .setUiTheme(runtimeConfig)
            .build()
    }
}